drivers/timer/apic_timer.c

```diff
@@ -192,20 +192,18 @@ uint32_t sys_clock_elapsed(void)
 
 #ifdef CONFIG_APIC_TIMER_TSC
 
-uint32_t sys_clock_cycle_get_32(void)
+uint64_t sys_clock_cycle_get_64(void)
 {
 	uint64_t tsc = z_tsc_read();
-	uint32_t cycles;
 
-	cycles = (tsc * CONFIG_APIC_TIMER_TSC_M) / CONFIG_APIC_TIMER_TSC_N;
-	return cycles;
+	return (tsc * CONFIG_APIC_TIMER_TSC_M) / CONFIG_APIC_TIMER_TSC_N;
```

Comment:
The kconfig docs don't make it clear, but these values are 32 bit numbers from the 0x15 leaf of CPUID, they won't overflow in 64 bits within reasonable runtimes (literally the same subject came up last week in the context of the apic_tsc PR that added a runtime query for the clock rate).

But here this is backwards anyway, the timer's clock should clearly be the value set in SYS_CLOCK_HW_CYCLES_PER_SEC.  What this code is doing is back-converting it to the "reference cycle clock" rate at runtime, which is useless because nothing else in the system cares about that clock.  Whether that should be fixed in this PR is an open question, but if we're going to be stuck with this driver we might as well put on a coat of paint...

Reply:
Ah fair enough, if tsc is a 32 bit value, and M is relatively small then indeed the concern isn't really relevant. That's not obvious I have to say from reading this, and presumably tsc could be a 64bit value on x86_64?

Reply:
CONFIG_APIC_TIMER_TSC_M couldn't be large. As @andyross said it's just a CPU setting to represent clock ratio.